### PR TITLE
Add auth-backend provider for AWS ALB OIDC reverse proxy

### PR DESCRIPTION
**Pull request by backjo**
## Hey, I just made a Pull Request!

This PR adds support for a new authentication provider - an AWS ALB OIDC reverse proxy. Since all authentication and authorization is managed upstream of Backstage, it only implements the refresh method. 

The change needing the most scrutiny in this PR is the addition of a `IdentityResolver` function to AuthProviderFactoryOptions. This pushes the responsibility of mapping authentication information to an identity into the individual backstage applications. This is mostly done to prevent scenarios like #3535 , but also because we simply don't have a standard identifier that comes down on the JWT from the ALB - it's likely to vary depending on the OIDC provider (i.e. if you're using the Microsoft Azure AD oidc provider, the sub claim will give a unique ID, but it's not actually linkable to the user ID in AAD)

One notable component missing from this PR is a UI component that triggers the 'login'. We use something like this:
```jsx
class DummySignInComponent extends React.Component<SignInPageProps, any> {
    constructor(props: any) {
        super(props)
        fetch(`${window.location.origin  }/api/auth/awsalb/refresh`).then(data => data.json()).then((data) => {
            props.onResult({userId: data.backstageIdentity.id, profile: data.profile})
        })
    }

    render() {
        return (
            <div />
        );
    }
}
```
to integrate our UI with this processor. I'm open to including it in the PR, but I know from previous issues that Spotify has one internally as well - maybe there's opportunity there for re-use across providers?

#### :heavy_check_mark: Checklist

- [X] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [X] Tests for new functionality and regression tests for bug fixes
- [X] Screenshots attached (for UI changes)

### CODE REVIEW

**Comment by freben** [comment]
Noting that it sounds like the `IdentityResolver` has a potential interesting overlap with https://github.com/backstage/backstage/issues/4089

File: plugins/auth-backend/src/providers/types.ts
Comment (Rugvip) [inline plugins/auth-backend/src/providers/types.ts:115]: Wondering if we can call this experimental for now. We have plenty of plans here, as freben mentioned, but don't wanna have to block this PR on getting the interface perfect.

**Review by Rugvip — COMMENTED** [review]
Thank you!

A good precedent for LB auth patterns, and thank you for porting to jose v3 :grin:

Regarding `DummySignInComponent` that's definitely a way to do it. We do likely want to add a core component for that use-case though, but prolly to support both single-provider and reverse-proxy auth

File: plugins/auth-backend/src/identity/TokenFactory.test.ts
Comment (freben) [inline plugins/auth-backend/src/identity/TokenFactory.test.ts:105]: At this point one might as well do `Number(payload.iat)` instead

File: plugins/auth-backend/src/providers/aws-alb/provider.ts
Comment (freben) [inline plugins/auth-backend/src/providers/aws-alb/provider.ts:106]: are these never expired or revoked?

File: plugins/auth-backend/package.json
Comment (OrkoHunter) [reply plugins/auth-backend/package.json:27]: I wonder if something like [this](https://github.com/backstage/backstage/blob/b2b9595ad4f70c10d09a8451fad89b7e5e98bc2c/packages/techdocs-common/package.json#L68-L73) can help jest find these packages -
```
  "jest": {
    "roots": [
      ".."  # Or something else
    ]
  }

```

File: plugins/auth-backend/src/identity/TokenFactory.ts
Comment (backjo) [reply plugins/auth-backend/src/identity/TokenFactory.ts:132]: Ah - this was not intended! Let me switch it back to HS256

File: plugins/auth-backend/package.json
Comment (backjo) [reply plugins/auth-backend/package.json:27]: Yeah, this was painful - see https://github.com/panva/jose/discussions/105 for context.

I suppose we could do the wildcard suggested there -> 
```text
"^jose/(.*)$": "<rootDir>/../node_modules/jose/dist/node/cjs/$1"
```

File: plugins/auth-backend/src/providers/types.ts
Comment (backjo) [reply plugins/auth-backend/src/providers/types.ts:115]: Yes, totally - this should be replaced with what is decided in #4089  - this is just a placeholder

Comment (backjo) [reply plugins/auth-backend/src/providers/types.ts]: I changed the type name here to ExperimentalIdentityResolver and added a comment in the type documentation calling out that it almost certain to change in the future.

File: plugins/auth-backend/src/providers/aws-alb/provider.ts
Comment (backjo) [reply plugins/auth-backend/src/providers/aws-alb/provider.ts]: Added node-cache and ttl'd these to 1 hour

File: plugins/auth-backend/src/identity/TokenFactory.ts
Comment (backjo) [reply plugins/auth-backend/src/identity/TokenFactory.ts]: I reworked this a bit and added a comment in the code about why this is necessary. Basically, TS doesn't differentiate between a missing property (no key) and a key that has an undefined value. So TSC is complaining here that it can't map this object to type AnyJWT (which extends from type Record<String, String>), because it can't validate that all keys have values of type string. In practice, the JOSE library is only setting keys/values on fields that it returns, and is not setting explicit "undefined" for these optional fields - but again, we have no way in TS to differentiate the two. 

https://github.com/microsoft/TypeScript/issues/13195

**Review by freben — APPROVED** [review]
Good from me!

File: plugins/auth-backend/src/identity/TokenFactory.ts
Comment (Rugvip) [inline plugins/auth-backend/src/identity/TokenFactory.ts:140]: Would prefer casts over using `@ts-ignore`, as it's easy to hide other type errors by mistake. I think in this case it might be enough to use

```ts
const storedJwk = {
...
} as AnyJWK
```

File: plugins/auth-backend/src/providers/aws-alb/provider.ts
Comment (Rugvip) [inline plugins/auth-backend/src/providers/aws-alb/provider.ts:104]: Let's not add `r2` and stick to existing packages instead, in this case simply use `fetch` via `cross-fetch` or `node-fetch`

File: plugins/auth-backend/package.json
Comment (Rugvip) [inline plugins/auth-backend/package.json:21]: 👍

**Comment by mfrinnstrom** [comment]
Looking forward to trying this out. We just configured our ALB yesterday 🙂

**Review by Rugvip — APPROVED** [review]
Thanks! Let's :shipit: and give it a spin :grin:

**Comment by mfrinnstrom** [comment]
@backjo sorry to bother you but I'm trying this out and I get a weird error:
```
FetchError: invalid json response body at https://public-keys.auth.elb.eu-west-1.amazonaws.com/<actaul-key-id> reason: Unexpected number in JSON at position 1
at /usr/src/app/node_modules/node-fetch/lib/index.js:272:32
at processTicksAndRejections (internal/process/task_queues.js:93:5)
at async AwsAlbAuthProvider.getKey (/usr/src/app/node_modules/@backstage/plugin-auth-backend/dist/index.cjs.js:1450:21)
at async AwsAlbAuthProvider.refresh (/usr/src/app/node_modules/@backstage/plugin-auth-backend/dist/index.cjs.js:1427:21)
```
It seems like it is the following lines that causes this but I can't understand why. Did you come across this at any point and can you give me some pointers in that case?

https://github.com/backstage/backstage/blob/789de4b164e5ec4e4b3580d263f9d6a129f680a1/plugins/auth-backend/src/providers/aws-alb/provider.ts#L103-L105

I'm deploying in `eu-west-1` so that should be ok.

**Comment by freben** [comment]
@mfrinnstrom what is the response if you query the URL manually? Including the content-type response header

**Comment by freben** [comment]
@mfrinnstrom what is the response if you query the URL manually? Including the content-type response header

**Comment by backjo** [comment]
@mfrinnstrom this should be fixed in #4224, which is not published yet.

**Comment by freben** [comment]
Ahh right i forgot about that one 😅

**Comment by mfrinnstrom** [comment]
@backjo thank you. I will try it with a nightly then and see if that helps.

**Comment by mfrinnstrom** [comment]
@backjo sorry to bother you here again (let me know if I should open up a support ticket instead).

I tried it with a nightly (`0.0.0-nightly-20210262290`) and the previous error went away. Instead I now get the following error:
```
TypeError: key must be an instance of a key instantiated by JWK.asKey, a valid JWK.asKey input, or a JWKS.KeyStore instance
at module.exports (/usr/src/app/node_modules/jose/lib/help/get_key.js:33:11)
at Object.module.exports [as verify] (/usr/src/app/node_modules/jose/lib/jwt/verify.js:229:9)
at AwsAlbAuthProvider.refresh (/usr/src/app/node_modules/@backstage/plugin-auth-backend/dist/index.cjs.js:1428:34)
at runMicrotasks (<anonymous>)\n at processTicksAndRejections (internal/process/task_queues.js:93:5)
```
The key is fetched from `https://public-keys.auth.elb.eu-west-1.amazonaws.com/a6e9048f-9d00-4690-b4a8-7989cd7ab6c8` and that seems to work as expected now.

From the error it seems it might have to do with how the keys are instantiated. I'm thinking of this line here: https://github.com/backstage/backstage/blob/789de4b164e5ec4e4b3580d263f9d6a129f680a1/plugins/auth-backend/src/providers/aws-alb/provider.ts#L106
I don't have enough knowledge about the libraries to be certain of that though. I would appreciate any pointers in getting this to work.

**Comment by backjo** [comment]
@mfrinnstrom mind sharing which version of `jose` is getting pulled in? The stack trace you shared seems to suggest 2.x, which I wouldn't expect to see here.

**Comment by backjo** [comment]
oh, ignore that last comment - I didn't realize the jose dependency change got reverted.

**Comment by backjo** [comment]
let me test with the #4184 changes and see what's up

**Comment by mfrinnstrom** [comment]
@backjo In my local environment I get `1.28.0` pulled in. I can see a reference to `2.0.3` in `yarn.lock` though.

**Comment by backjo** [comment]
@mfrinnstrom the only issue I ran into running the latest version was the issuer verification, which has a fix in #4266

**Comment by mfrinnstrom** [comment]
@backjo Ok, thanks for looking into it. I will do some more checking and see if I can make it work then. I'm running with an app created with `create-app`.

I'm configuring the `auth` plugin like this in the backend if that might have something to do with it. Not all parts are filled in yet 🙂 
```typescript
export default async function createPlugin({
  logger,
  database,
  config,
  discovery,
}: PluginEnvironment) {
  const identityResolver = (
    payload: object,
    // catalogApi: CatalogApi,
  ): Promise<AuthResponse<any>> => {
    return Promise.resolve({
      providerInfo: {},
      profile: {
        email: '',
        displayName: '',
        picture: '',
      },
      backstageIdentity: {
        id: '',
      }
    });
  };
  const providerFactories = {
    awsalb: (options: AuthProviderFactoryOptions) => defaultAuthProviderFactories.awsalb({...options, identityResolver})
  }
  return await createRouter({ logger, config, database, discovery, providerFactories });
}
```
It seems a bit cumbersome but that was the only way I could find to add my own `identityResolver`. I also have a dummy sign in page based on your example.

**Comment by backjo** [comment]
@mfrinnstrom we're using it this exact way too right now. I agree that it's cumbersome, but I think it makes sense to wait until @freben 's #4089 lands so that the interface exposed for the identity resolver is standard. 

BTW, I figured out the key issue. Turns out the issue was reading the key back from the cache - so I thought it was working because the first request (to load backstage) always worked! Fix coming for that shortly

**Comment by backjo** [comment]
@mfrinnstrom - fix is in #4268 . Really appreciate your patience and help getting this to work :)

**Comment by mfrinnstrom** [comment]
@backjo no problem, thanks for fixing it 🙂 I hope that this can make it into the next release, which is tomorrow I guess.

Regarding @freben 's #4089 I agree that it makes sense to wait for that one. I was just wondering if I had missed an easier way of doing the config.

**Comment by mfrinnstrom** [comment]
@backjo another question if I may. How do handle this during local development? Do you activate the `awsalb` provider based on some property?

**Comment by backjo** [comment]
yeah, just an enabled/disable flag on the configuration, where the awsalb flag is set to disabled on local development.

**Comment by regicsolutions** [comment]
@backjo Is there any documentation on getting this up and running? was looking into a way to call lambda functions from backstage.

**Pull request merged**
